agent/consul/intention_endpoint.go

```diff
@@ -318,180 +243,132 @@ func (s *Intention) computeApplyChangesLegacyUpdate(
-	// validation, if we do them here too we can generate error messages that
-	// make more sense for legacy edits.
 	if !ixn.CanWrite(authz) {
 		// todo(kit) Migrate intention access denial logging over to audit logging when we implement it
 		s.logger.Warn("Deletion operation on intention denied due to ACLs", "intention", args.Intention.ID, "accessorID", accessorID)
-		return "", nil, acl.ErrPermissionDenied
+		return nil, acl.ErrPermissionDenied
 	}
 
-	upsertEntry := prevEntry.Clone()
-
-	deleted := upsertEntry.DeleteSourceByLegacyID(args.Intention.ID)
-	if !deleted {
-		return "", nil, fmt.Errorf("Cannot delete non-existent intention: '%s'", args.Intention.ID)
-	}
-
-	if upsertEntry == nil || len(upsertEntry.Sources) == 0 {
-		return structs.ConfigEntryDelete, &structs.ServiceIntentionsConfigEntry{
-			Kind:           structs.ServiceIntentions,
-			Name:           prevEntry.Name,
-			EnterpriseMeta: prevEntry.EnterpriseMeta,
-		}, nil
-	}
-
-	return structs.ConfigEntryUpsertCAS, upsertEntry, nil
+	return &structs.IntentionMutation{
+		ID: args.Intention.ID,
+	}, nil
 }
 
 func (s *Intention) computeApplyChangesDelete(
+	accessorID string,
+	authz acl.Authorizer,
 	entMeta *structs.EnterpriseMeta,
 	args *structs.IntentionRequest,
-) (structs.ConfigEntryOp, *structs.ServiceIntentionsConfigEntry, error) {
+) (*structs.IntentionMutation, error) {
 	args.Intention.DefaultNamespaces(entMeta)
 
-	prevEntry, err := s.getServiceIntentionsConfigEntry(args.Intention.DestinationName, args.Intention.DestinationEnterpriseMeta())
-	if err != nil {
-		return "", nil, err
-	}
-
-	if prevEntry == nil {
-		return "", nil, nil // no op means no-op
-	}
-
-	// NOTE: validation errors may be misleading!
-
-	upsertEntry := prevEntry.Clone()
-
-	sn := args.Intention.SourceServiceName()
-
-	deleted := upsertEntry.DeleteSourceByName(sn)
-	if !deleted {
-		return "", nil, nil // no op means no-op
-	}
-
-	if upsertEntry == nil || len(upsertEntry.Sources) == 0 {
-		return structs.ConfigEntryDelete, &structs.ServiceIntentionsConfigEntry{
-			Kind:           structs.ServiceIntentions,
-			Name:           prevEntry.Name,
-			EnterpriseMeta: prevEntry.EnterpriseMeta,
-		}, nil
+	if !args.Intention.CanWrite(authz) {
+		// todo(kit) Migrate intention access denial logging over to audit logging when we implement it
+		s.logger.Warn("Intention delete denied due to ACLs", "accessorID", accessorID)
+		return nil, acl.ErrPermissionDenied
 	}
 
-	return structs.ConfigEntryUpsertCAS, upsertEntry, nil
-}
-
-func checkLegacyIntentionApplyAllowed(prevEntry *structs.ServiceIntentionsConfigEntry) error {
-	if prevEntry == nil {
-		return nil
+	// Pre-flight to avoid pointless raft operations.
+	_, _, ixn, err := s.srv.fsm.State().IntentionGetExact(nil, args.Intention.ToExact())
+	if err != nil {
+		return nil, fmt.Errorf("Intention lookup failed: %v", err)
 	}
-	if prevEntry.LegacyIDFieldsAreAllSet() {
-		return nil
+	if ixn == nil {
+		return nil, nil
```

Comment:
How come this doesn't return the "can't delete non-existent intention" error?

Reply:
The new endpoint variations that operate by name instead of by ID did not retain the bizarre behavior of their predecessors. Deletes the new way are idempotent so you can re-delete an infinite number of times and it'll say "sure!"

I suspect partly the ID-based deletes had to work the way they did because you couldn't enforce ACLs unless you could turn the ID into source+destination pairs, and if it doesn't exist it can't tell you if something is possible or not possible.

In the name-based endpoints all of the information necessary for the ACL check is part of the request.